src/zjs_net.c

```diff
@@ -0,0 +1,1111 @@
+        if (!handle->timer_started) {
+            // time has not been started
+            k_timer_init(&handle->timer, socket_timeout_callback, NULL);
+            handle->timer_started = 1;
+        }
+        k_timer_start(&handle->timer, time, time);
+        if (handle->tcp_timeout_id == -1) {
+            handle->tcp_timeout_id = zjs_add_c_callback(handle, tcp_c_timeout_callback);
+        }
+        DBG_PRINT("starting socket timeout: %u\n", time);
+    } else if (handle->timer_started) {
+        DBG_PRINT("stoping socket timeout\n");
+        k_timer_stop(&handle->timer);
+    }
+    handle->timeout = time;
+}
+
+static void tcp_c_callback(void *h, const void *args)
+{
+    sock_handle_t *handle = (sock_handle_t *)h;
+    if (handle) {
+        // find length of unconsumed data in read buffer
+        uint32_t len = handle->wptr - handle->rptr;
+        zjs_buffer_t *zbuf;
+        ZVAL data_buf = zjs_buffer_create(len, &zbuf);
+        // copy data from read buffer
+        memcpy(zbuf->buffer, handle->rptr, len);
+        handle->rptr = handle->wptr = handle->rbuf;
+
+        zjs_trigger_event(handle->socket, "data", &data_buf, 1, NULL, NULL);
+
+        zjs_remove_callback(handle->tcp_read_id);
+    } else {
+        ERR_PRINT("handle is NULL\n");
+    }
+}
+
+static void post_server_closed(void *handle)
+{
+    net_handle_t *h = (net_handle_t *)handle;
+    if (h) {
+        DBG_PRINT("closing server\n");
+        net_context_put(h->tcp_sock);
+        zjs_free(h);
+    }
+}
+
+static void post_closed(void *handle)
+{
+    sock_handle_t *h = (sock_handle_t *)handle;
+    if (h) {
+        net_handle_t *net = h->handle;
+        sock_handle_t *cur = opened_sockets;
+        if (cur->next == NULL) {
+            opened_sockets = NULL;
+            net_context_put(cur->tcp_sock);
+            jerry_release_value(cur->socket);
+            zjs_free(cur->rbuf);
+            zjs_free(cur);
+            DBG_PRINT("Freed socket: opened_sockets=%p\n", opened_sockets);
+        } else {
+            while (cur->next) {
+                if (cur->next == h) {
+                    cur->next = cur->next->next;
+                    net_context_put(cur->tcp_sock);
+                    jerry_release_value(cur->socket);
+                    zjs_free(cur->rbuf);
+                    zjs_free(cur);
+                    DBG_PRINT("Freed socket: opened_sockets=%p\n", opened_sockets);
+                }
+            }
+        }
+        if (net) {
+            if (net->listening == 0 && opened_sockets == NULL) {
+                // no more sockets open and not listening, close server
+                zjs_trigger_event(net->server, "close", NULL, 0, post_server_closed, net);
+                DBG_PRINT("server signaled to close\n");
+            }
+        }
+    }
+}
+
+static void tcp_received(struct net_context *context,
+                         struct net_buf *buf,
+                         int status,
+                         void *user_data)
+{
+
```

Comment:
remove blank line

And check status; if 0, success; if <0 you're supposed to trigger 'error' event.